### DISCUSSION TITLE
URL filtering in Moodle

### DISCUSSION THREAD

**Discussion by cienciascontic** [SafeExamBrowser/seb-win-refactoring#137] [General]
Hi! I've configured my Moodle quiz to enable access to Rstudio using the following expression: 'rstudio.cloud/*' but when a student tries to open the quiz, SEB displays an alter windows saying: "Access to "" is not allowed according to the current configuration". Am I doing something wrong?

I want students to get access to Rstudio while they're doing the exam providing the link in one of the questions, but I don't want to them to access any other webpage except Rstudio (since Twitter link is displayed at the bottom of Rstudio page will be easy for them to sign in into their twitter account and exchange answers). That's why I decided to use the filtering feature which appears to not work properly.
Please help.
Thanks a lot.

**Comment by dbuechel** [comment]
You can either check the client log file (to be found under `%LocalAppData%\SafeExamBrowser\Logs`) after a session to see exactly which URLs are being blocked, or you can enable the real-time application log within SEB to immediately see whether a URL is being blocked or not. Enable that URLs are visible in the log (see the first picture below) and for the latter, allow access to the application log during runtime (see the second picture).

![grafik](https://user-images.githubusercontent.com/28347927/127630161-30644f0f-4723-42f2-a870-4b7f6a5c5ede.png)
![grafik](https://user-images.githubusercontent.com/28347927/127630209-ac76a497-0e5f-483a-bb11-b642061480ab.png)

**Reply by cienciascontic** [reply]
Thanks for your answer. I'm here providing a [link to the client log file](https://drive.google.com/file/d/1L7WYlvrMwmYMCzbZjx5SxaaSg7yFu0Pn/view?usp=sharing), but the problem is that the Moodle quiz does not open, students always get the same error message: ""Access to '' is not allowed according to the current configuration".

**Reply by danschlet** [reply]
You have to add URL filter rules which allow access to your Moodle server and in case you’re using a single-sign on system on another domain, you also need to allow access to that.

SEB only allows access to the exact URL of the SEB setting Start URL, which is the exact Moodle quiz URL. But you need to allow access to the whole domain/path of your Moodle instance, like https://example.org/MoodlePath/*

**Reply by cienciascontic** [reply]
Ok, thanks. I've changed that including my Moodle path within the filter rules but still being unable to access the quiz as a student, I still get the same error.
Here's how I'm configuring the URL filter rules for my quiz:
![image](https://user-images.githubusercontent.com/6078186/127660651-9f5611ca-7ad8-44e5-b12f-41f6a18e13b2.png)
I understand that if I use the expression 'example.com/*' any URL paths after the / will be accessible. Am I right? Thanks a ton.

**Reply by danschlet** [reply]
> I understand that if I use the expression 'example.com/*' any URL paths after the / will be accessible. Am I right? 

Yes, that's correct. As @dbuechel mentioned, you need to find out in the log file which URLs got blocked, maybe your server or network (proxy etc.) returns another URL than the one mentioned. I don't see that in the log file you provided, maybe it's not the right one or @dbuechel can help further.

You can also use the Mac version of SEB and activate the [Teach SEB allowed/blocked URLs](https://safeexambrowser.org/macosx/mac_usermanual_en.html#NetworkPaneFilterSection) feature, which will interactively show the blocked URLs and allow to load them to create the correct filter rules.

**Reply by cienciascontic** [reply]
Thanks for your help @danschlet and @dbuechel  I've done what you suggested, activating the settings on the SEB config utility but no URLs are displayed as blocked within the log file which I'm here sharing: [https://drive.google.com/file/d/1staJam2hyoBAqbUdn6Dj3sMK5rydqpN3/view?usp=sharing](https://drive.google.com/file/d/1staJam2hyoBAqbUdn6Dj3sMK5rydqpN3/view?usp=sharing)
What I also see that SEB is blocking is the stylesheet (???). Thanks again for your support.

**Reply by dbuechel** [reply]
Yes, there are requests being blocked, e.g.:

```
2021-07-30 12:10:34.705 [19] - INFO: [Browser Instance #1] Blocked content request (Stylesheet, LinkClicked).
2021-07-30 12:10:34.730 [19] - INFO: [Browser Instance #1] Blocked content request (Image, LinkClicked).
2021-07-30 12:10:34.734 [19] - INFO: [Browser Instance #1] Blocked content request (Script, LinkClicked).
2021-07-30 12:10:38.807 [18] - INFO: [RequestHandler #1] Blocked main request (MainFrame, FormSubmit).
```

As explained above, you need to activated the logging of URLs in the configuration tool under tab "Browser" -> "Show URLs" to see exactly which request is being blocked.

**Reply by cienciascontic** [reply]
@dbuechel Thanks for your patience. Trust me that I already configured that as you suggested, but the problem is that the quiz doesn't start, I never get into the Moodle quiz. When I click on the button to start the quiz I get the Page Blocked message and no URL appears in any part of the SEB window. This couldn't be so hard... It must be something that is not working from the Moodle side with this configuration. Thanks again.

**Reply by danschlet** [reply]
The problem is that you would need to change the settings @dbuechel mentioned in the configuration which Moodle is providing when you click the "Launch Safe Exam Browser" button. You could do that by using the option in the Safe Exam Browser settings in the Moodle quiz to "upload my own config". Or you could download the settings Moodle provides for the quiz and then edit those and start SEB manually with the edited file (just open/double click it).

Also you could file a feature request/improvement issue at moodle.org to provide an option to automatically allow the whole Moodle server when the URL filter rules are used for a quiz. I agree that this could be easier to use, but it has to be improved in Moodle, as SEB cannot "know" exactly which URLs should be allowed. In the SEB documentation we mention that the URL filter feature is complicated to use correctly, we only recommend this to administrators which exactly know what they are doing and can analyze such problems you are observing.

**Reply by cienciascontic** [reply]
@danschlet I already did that and SEB doesn't show any URLs. Try your self with these credentials (don't mind, it's a fake student):

[Click here to enter into Moodle](https://campusvirtual.udesa.edu.ar)

username: linmat
password: SEB123456*

Enter this course: 
![image](https://user-images.githubusercontent.com/6078186/127708614-5dc56893-cd86-4ef6-8763-25a3b1c20c54.png)

and then select the "**SEB test Moodle quiz**" exam.

Let me know then if you see which URL is being blocked.
Kind regards.

**Reply by danschlet** [reply]
The issue is caused by the '...' quotation marks around the URL filter rules. Did you enter those yourself? If not, then maybe there's some bug caused by the different language or locale settings.

Without quotes at least the general rule allowing the whole server works. Your configuration file contains various Moodle URLs (I didn't try to remove the quotes there), did you enter all of those yourself? I don't really know if Moodle already adds some allow rules for its own server (as we didn't develop the Moodle integration and I personally never tried the URL rule setting). In any case it is not easy to allow only specific parts of Moodle to be accessible, because it uses various paths and php scripts with complex URLs (and some are not visible in a regular browser if you would try to find them out manually).

**Reply by cienciascontic** [reply]
I can't believe that was the problem.... oh my gosh.... I took that from SEB's manual, but too literally:
![image](https://user-images.githubusercontent.com/6078186/127715157-e4720d9a-b952-4fef-b9a7-1da5b142aca0.png)
Thanks a ton for your help.
Now the filter is working perfectly.
Warm regards.
